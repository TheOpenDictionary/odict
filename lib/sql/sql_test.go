package sql

import (
	"math/rand"
	"testing"

	"github.com/TheOpenDictionary/odict/lib/core"
	"github.com/TheOpenDictionary/odict/lib/test"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/google/uuid"
	"github.com/stretchr/testify/assert"
)

func TestDictionaryToSQL(t *testing.T) {
	_, err := core.CompileDictionary("../../examples/example1.xml", "../../examples/example1.odict")

	assert.Equal(t, nil, err)

	dict, err := core.ReadDictionary("../../examples/example1.odict")

	assert.Equal(t, nil, err)

	dict_r := dict.AsRepresentable()

	assert.Equal(t, nil, err)

	uuid.SetRand(rand.New(rand.NewSource(1)))

	sql, err := DictionaryToSQL(dict_r, "postgres", true)

	assert.Equal(t, nil, err)

	snaps.MatchSnapshot(t, sql)

	test.CleanupTest()
}
