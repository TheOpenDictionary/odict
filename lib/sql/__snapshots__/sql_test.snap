
[TestDictionaryToSQL - 1]
CREATE TABLE dictionaries (
    id UUID NOT NULL
    ,name TEXT NOT NULL

    ,CONSTRAINT dictionaries_id_pkey PRIMARY KEY (id)
);

CREATE TABLE entries (
    id UUID NOT NULL
    ,created_at TIMESTAMPTZ DEFAULT CURRENT_TIMESTAMP
    ,updated_at TIMESTAMPTZ
    ,term TEXT NOT NULL
    ,dictionary_id UUID

    ,CONSTRAINT entries_id_pkey PRIMARY KEY (id)
);

CREATE TABLE etymologies (
    id UUID NOT NULL
    ,description TEXT
    ,entry_id UUID NOT NULL

    ,CONSTRAINT etymologies_id_pkey PRIMARY KEY (id)
);

CREATE TABLE senses (
    id UUID NOT NULL
    ,pos TEXT DEFAULT 'unknown'
    ,etymology_id UUID NOT NULL

    ,CONSTRAINT senses_id_pkey PRIMARY KEY (id)
);

CREATE TABLE groups (
    id UUID NOT NULL
    ,description TEXT
    ,sense_id UUID NOT NULL

    ,CONSTRAINT groups_id_pkey PRIMARY KEY (id)
);

CREATE TABLE definitions (
    id UUID NOT NULL
    ,value TEXT NOT NULL
    ,sense_id UUID
    ,group_id UUID

    ,CONSTRAINT definitions_id_pkey PRIMARY KEY (id)
);

CREATE TABLE notes (
    id UUID NOT NULL
    ,value TEXT
    ,definition_id UUID NOT NULL

    ,CONSTRAINT notes_id_pkey PRIMARY KEY (id)
);

CREATE TABLE examples (
    id TEXT NOT NULL
    ,text TEXT NOT NULL
    ,definition_id UUID
    ,note_id UUID

    ,CONSTRAINT examples_id_pkey PRIMARY KEY (id)
);

ALTER TABLE entries ADD CONSTRAINT entries_dictionary_id_fkey FOREIGN KEY (dictionary_id) REFERENCES dictionaries (id);

ALTER TABLE etymologies ADD CONSTRAINT etymologies_entry_id_fkey FOREIGN KEY (entry_id) REFERENCES entries (id);

ALTER TABLE senses ADD CONSTRAINT senses_etymology_id_fkey FOREIGN KEY (etymology_id) REFERENCES etymologies (id);

ALTER TABLE groups ADD CONSTRAINT groups_sense_id_fkey FOREIGN KEY (sense_id) REFERENCES senses (id);

ALTER TABLE definitions ADD CONSTRAINT definitions_sense_id_fkey FOREIGN KEY (sense_id) REFERENCES senses (id);

ALTER TABLE definitions ADD CONSTRAINT definitions_group_id_fkey FOREIGN KEY (group_id) REFERENCES groups (id);

ALTER TABLE notes ADD CONSTRAINT notes_definition_id_fkey FOREIGN KEY (definition_id) REFERENCES definitions (id);

ALTER TABLE examples ADD CONSTRAINT examples_definition_id_fkey FOREIGN KEY (definition_id) REFERENCES definitions (id);

ALTER TABLE examples ADD CONSTRAINT examples_note_id_fkey FOREIGN KEY (note_id) REFERENCES notes (id);
;

INSERT INTO dictionaries (name, id) VALUES ('Example Dictionary 1', '52fdfc07-2182-454f-963f-5f0f9a621d72');
INSERT INTO entries (id, term, dictionary_id) VALUES ('9566c74d-1003-4c4d-bbbb-0407d1e2c649', 'cat', '52fdfc07-2182-454f-963f-5f0f9a621d72');
INSERT INTO etymologies (id, description, entry_id) VALUES ('81855ad8-681d-4d86-91e9-1e00167939cb', 'Latin root', '9566c74d-1003-4c4d-bbbb-0407d1e2c649');
INSERT INTO senses (id, etymology_id, pos) VALUES ('6694d2c4-22ac-4208-a007-2939487f6999', '81855ad8-681d-4d86-91e9-1e00167939cb', 'n');
INSERT INTO definitions (id, value, sense_id) VALUES ('eb9d18a4-4784-445d-87f3-c67cf22746e9', 'a cat', '6694d2c4-22ac-4208-a007-2939487f6999');
INSERT INTO notes (id, value, definition_id) VALUES ('95af5a25-3679-41ba-a2ff-6cd471c483f1', 'Some definition note', 'eb9d18a4-4784-445d-87f3-c67cf22746e9');
INSERT INTO examples (id, text, note_id) VALUES ('5fb90bad-b37c-4821-b6d9-5526a41a9504', 'Some example', '95af5a25-3679-41ba-a2ff-6cd471c483f1');
INSERT INTO examples (id, text, definition_id) VALUES ('680b4e7c-8b76-4a1b-9d49-d4955c848621', 'There goes a cat!', 'eb9d18a4-4784-445d-87f3-c67cf22746e9');
INSERT INTO entries (id, term, dictionary_id) VALUES ('6325253f-ec73-4dd7-a9e2-8bf921119c16', 'dog', '52fdfc07-2182-454f-963f-5f0f9a621d72');
INSERT INTO etymologies (id, description, entry_id) VALUES ('0f070244-8615-4bda-8831-3f6a8eb668d2', 'Latin root', '6325253f-ec73-4dd7-a9e2-8bf921119c16');
INSERT INTO senses (id, etymology_id, pos) VALUES ('0bf50598-7592-4e66-8a5b-df2c7fc48445', '0f070244-8615-4bda-8831-3f6a8eb668d2', 'un');
INSERT INTO definitions (id, value, sense_id) VALUES ('92d2572b-cd06-48d2-96c5-2f5054e2d083', 'a dog', '0bf50598-7592-4e66-8a5b-df2c7fc48445');
INSERT INTO entries (id, term, dictionary_id) VALUES ('6bf84c71-74cb-4476-b64c-c3dbd968b0f7', 'poo', '52fdfc07-2182-454f-963f-5f0f9a621d72');
INSERT INTO etymologies (id, description, entry_id) VALUES ('172ed857-94bb-458b-8c3b-525da1786f9f', 'Latin root', '6bf84c71-74cb-4476-b64c-c3dbd968b0f7');
INSERT INTO senses (id, etymology_id, pos) VALUES ('ff094279-db19-44eb-97a1-9d0f7bbacbe0', '172ed857-94bb-458b-8c3b-525da1786f9f', 'un');
INSERT INTO groups (id, description, sense_id) VALUES ('255aa5b7-d44b-4c40-b84c-892b9bffd436', 'A number of verb senses', 'ff094279-db19-44eb-97a1-9d0f7bbacbe0');
INSERT INTO definitions (id, value) VALUES ('29b0223b-eea5-44f7-8391-f445d15afd42', 'crap, shit');
INSERT INTO entries (id, term, dictionary_id) VALUES ('94040374-f692-4b98-8bf8-713f8d962d7c', 'ran', '52fdfc07-2182-454f-963f-5f0f9a621d72');
INSERT INTO entries (id, term, dictionary_id) VALUES ('8d019192-c242-44e2-8afc-cae3a61fb586', 'run', '52fdfc07-2182-454f-963f-5f0f9a621d72');
INSERT INTO etymologies (id, description, entry_id) VALUES ('b14323a6-bc8f-4e7d-b1d9-29333ff99393', 'Latin root', '8d019192-c242-44e2-8afc-cae3a61fb586');
INSERT INTO senses (id, etymology_id, pos) VALUES ('3bea6f5b-3af6-4e03-b436-6c4719e43a1b', 'b14323a6-bc8f-4e7d-b1d9-29333ff99393', 'n');
INSERT INTO definitions (id, value, sense_id) VALUES ('067d89bc-7f01-41f5-b398-1659a44ff17a', '(vertebrates) To move swiftly', '3bea6f5b-3af6-4e03-b436-6c4719e43a1b');
INSERT INTO examples (id, text, definition_id) VALUES ('4c7215a3-b539-4b1e-9849-c6077dbb5722', 'The dog runs after the cat.', '067d89bc-7f01-41f5-b398-1659a44ff17a');
INSERT INTO examples (id, text, definition_id) VALUES ('f5717a28-9a26-4f97-a479-81998ebea89c', 'The horse ran away.', '067d89bc-7f01-41f5-b398-1659a44ff17a');
INSERT INTO definitions (id, value, sense_id) VALUES ('0b4b3739-7011-4e82-ad6f-4125c8fa7311', 'Act or instance of hurrying (to or from a place) (not necessarily by foot); dash or errand, trip.', '3bea6f5b-3af6-4e03-b436-6c4719e43a1b');
INSERT INTO definitions (id, value, sense_id) VALUES ('e4d7defa-922d-4ae7-b866-67f7e936cd4f', 'A pleasure trip.', '3bea6f5b-3af6-4e03-b436-6c4719e43a1b');
INSERT INTO definitions (id, value, sense_id) VALUES ('24abf7df-866b-4a56-8383-67ad6145de1e', 'Flight, instance or period of fleeing.', '3bea6f5b-3af6-4e03-b436-6c4719e43a1b');
INSERT INTO definitions (id, value, sense_id) VALUES ('e8f4a8b0-993e-4df8-883a-0ad8be9c3978', 'Migration (of fish).', '3bea6f5b-3af6-4e03-b436-6c4719e43a1b');
INSERT INTO definitions (id, value, sense_id) VALUES ('b04883e5-6a15-4a8d-a563-afa467d49dec', 'A group of fish that migrate, or ascend a river for the purpose of spawning.', '3bea6f5b-3af6-4e03-b436-6c4719e43a1b');
INSERT INTO senses (id, etymology_id, pos) VALUES ('6a40e9a1-d007-4033-8282-3061bdd0eaa5', 'b14323a6-bc8f-4e7d-b1d9-29333ff99393', 'v');
INSERT INTO groups (id, description, sense_id) VALUES ('9f8e4da6-4301-4522-8d0b-29688b734b8e', 'A number of verb senses', '6a40e9a1-d007-4033-8282-3061bdd0eaa5');
INSERT INTO definitions (id, value) VALUES ('a0f3ca99-36e8-461f-90d7-7c96ea80a7a6', '(vertebrates) To move swiftly (figuratively)');
INSERT INTO examples (id, text, definition_id) VALUES ('65f606f6-a63b-4f3d-bd25-67c18979e4d6', 'The dog runs after the cat.', 'a0f3ca99-36e8-461f-90d7-7c96ea80a7a6');
INSERT INTO examples (id, text, definition_id) VALUES ('0f26686d-9bf2-4b26-8901-ff354cde1607', 'The horse ran away.', 'a0f3ca99-36e8-461f-90d7-7c96ea80a7a6');
INSERT INTO definitions (id, value) VALUES ('ee294b39-f32b-4c78-a2ba-64f84ab43ca0', '(fluids) To flow.');
INSERT INTO definitions (id, value) VALUES ('c6e6b91c-1fd3-4e89-9043-4179d3af4491', '(nautical, of a vessel) To sail before the wind, in distinction from reaching or sailing close-hauled.');
INSERT INTO definitions (id, value) VALUES ('a369012d-b92d-484f-839d-1734ff571642', '(social) To carry out an activity.');
INSERT INTO definitions (id, value) VALUES ('8953bb68-65fc-492b-8c3a-17c9028be991', 'To extend or persist, statically or dynamically, through space or time.');
INSERT INTO definitions (id, value) VALUES ('4eb7649c-6c93-4780-8979-d1830356f2a5', '(transitive) To execute or carry out a plan, procedure or program.');
---
